### DISCUSSION TITLE
Supported Browser/JavaScript features for embeded web games

### DISCUSSION THREAD

**Discussion by mikekasprzak** [ludumdare/ludumdare#2172] [Ideas]
At the time of this writing, our embedded web games support these additional features.

* `requestFullScreen()`
* Performance features, via [cross-origin isolation](https://web.dev/cross-origin-isolation-guide/) (`SharedArrayBuffer`, high resolution timers)
* WebAssembly `wasm` files (`Content-Type: application/wasm`)

Leave a comment below if there's something we should enable (other than persistent storage and access to external websites).

**Comment by bcmpinc** [comment]
For cross-origin isolation to work and hence allow SharedArrayBuffer to be used, the html page containing the iframe needs to have the cross-origin isolation headers set as well. Would this be possible to add to submission pages, or would this break the website?

I've read some mentions that some parameters need to be added to the iframe as well, but I have not found reliable information on that.

For reference itch.io uses the following iframe tag:
`<iframe src="https://example.com/index.html" scrolling="no" mozallowfullscreen="true" id="game_drop" allowfullscreen="true" allowtransparency="true" webkitallowfullscreen="true" frameborder="0" msallowfullscreen="true" allow="autoplay; fullscreen *; geolocation; microphone; camera; midi; monetization; xr-spatial-tracking; gamepad; gyroscope; accelerometer; xr; cross-origin-isolated"></iframe>`
and adds these http-headers:
```
Cross-Origin-Embedder-Policy: credentialless
Cross-Origin-Opener-Policy: same-origin
```

Unfortunately, at this moment only Chrome and Edge support [credentialless](https://caniuse.com/mdn-http_headers_cross-origin-embedder-policy_credentialless). Firefox does not, but does ship experimental support that website owners can [opt-in](https://wiki.mozilla.org/Origin_Trials) for.

**Reply by seb776** [reply]
Hello,

Any news on this ?
I get the following errors while embedding Godot4 builds :/
The following features required to run Godot projects on the Web are missing:
Cross Origin Isolation - Check web server configuration (send correct headers)
SharedArrayBuffer - Check web server configuration (send correct headers)

Perhaps I'm missing something ?

**Reply by BartekDe** [reply]
@seb776 hey, did you find any workaround for this issue?

**Reply by seb776** [reply]
I finally published to itch.io with redirect links

**Reply by mikekasprzak** [reply]
I need to double check, but if I recall the headers are set correctly. The issue is that "correct" isn't necessarily correct for game hosting websites. Browsers only isolate JavaScript at the domain level, which is enough for 99.99% of websites. By design we set a policy that disables local storage, as technically that's the right thing to do for security. The problem is browsers also disable other features when you do this (and off hand I'm not sure what storage even has to do with it), and this is what's tripping up the Godot runtime. 

I was working on a patch that I had intended to upstream to Godot mainline, but having watched all the problems folks were having, it also made sense to change what we were doing to solve the issue: by sandboxing each game with a unique subdomain, we _can_ enable local storage. This was in progress, but when "the spam" problem got out of hand, I had to switch gears.

Alas that Godot patch of mine will probably never be finished and upstreamed, since practically few people will ever need it.

**Reply by seb776** [reply]
I think there should be a way to handle that as Itch.io can.
Actually this issue is preventing all Godot4 html5 builds so this will be a growing issue.
Perhaps I could check it out ?
After a quick research, I think the iframe is not crossOriginIsolated and it seems to miss the "allow-same-origin" value on the sandbox property of the iFrame.
This can be easily tested. It could have a side effect on other projects I think and that's probably why Itch.io commes with a "SharedArrayBuffer (Experimental)" toggle to activate it.

**Reply by samuelfine** [reply]
Just chiming in that I'm also experiencing this issue, and am a web developer who is available to assist and/or test if it'd be useful. Let me know if there's anything I can do to help! :)

**Comment by hybridherbst** [comment]
I'm using three.js with .glb files, which are loaded as blobs. Unfortunately it seems like the current CSP doesn't allow local blobs (?):
```
Refused to connect to 'blob:null/1b9f0e3c-be2e-4a66-a9bd-73ab6e316843' 
because it violates the following Content Security Policy directive: "default-src 'self' data:". 
Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.
```

My understanding is that the policy would need to be "default-src 'self' data: blob:" instead. Or is there another way to make this work?
